File: huffman.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
using namespace std;

struct Node {
    char ch;
    int val;
    Node *l;
    Node *r;
    Node(int val,char ch) : val(val),ch(ch), l(nullptr), r(nullptr) {}
};

class MinHeap {
    vector<Node*> heap;

    void heapifyup(int index) {
        while (index > 0) {
            int parentIdx = (index - 1) / 2;
            if (heap[index]->val < heap[parentIdx]->val) {
                swap(heap[index], heap[parentIdx]);
                index = parentIdx;
            } else {
                break;
            }
        }
    }

    void heapifydown(int index) {
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;
        int smallestIdx = index;

        if (leftChild < heap.size() && heap[leftChild]->val < heap[smallestIdx]->val) {
            smallestIdx = leftChild;
        }

        if (rightChild < heap.size() && heap[rightChild]->val < heap[smallestIdx]->val) {
            smallestIdx = rightChild;
        }

        if (smallestIdx != index) {
            swap(heap[index], heap[smallestIdx]);
            heapifydown(smallestIdx);
        }
    }

public:
    MinHeap() {}

    void insert(Node* node) {
        heap.push_back(node);
        heapifyup(heap.size() - 1);
    }

    Node* getMin() {
        if (heap.empty()) {
            cout << "Heap is Empty" << endl;
            return nullptr; 
        }
        return heap[0];
    }

    void removeMin() {
        if (heap.empty()) {
            cout << "Heap is empty" << endl;
            return; 
        }
        heap[0] = heap.back();
        heap.pop_back();
        heapifydown(0);
    }

    bool isEmpty() {
        return heap.empty();
    }

    int size() {
        return heap.size();
    }
};

void printCodes(unordered_map<char,string>&huffmanCodes){
    cout<<"Huffman Codes: "<<endl;
    for(const auto &pair:huffmanCodes){
        cout<<pair.first<<" : "<<pair.second<<endl;
    }
}

void generateCodes(Node* root,string str,unordered_map<char,string>&huffmanCodes){
    if(root==nullptr){
        return;
    }
    if(root->l==NULL && root->r==NULL){
        huffmanCodes[root->ch]=str;
    }
    generateCodes(root->l,str+"0",huffmanCodes);
    generateCodes(root->r,str+"1",huffmanCodes);
}

string encode(string str,unordered_map<char,string> &huffmanCodes){
    string encoded="";
    for(int i=0;i<str.size();i++){
        encoded+=huffmanCodes[str[i]];
    }
    return encoded;
}

void decode(string encoded,Node* root){
    string decoded="";
    Node*temp=root;
    for(int i=0;i<encoded.size();i++){
        
        if(encoded[i]=='0'){
            temp=temp->l;
        }else if(encoded[i]=='1'){
            temp=temp->r;
        }
        if(temp->l==NULL && temp->r==NULL){
            decoded+=temp->ch;
            temp=root;
        }
    }

    cout<<decoded<<endl;
}

Node* huffman(string str,char data[],int val[], int size) {
    MinHeap minHeap;

    // Insert all values as leaf nodes
    for (int i = 0; i < size; i++) {
        minHeap.insert(new Node(val[i],data[i]));
    }

    Node* root = nullptr;  // Track the root of the Huffman tree

    // Build Huffman Tree
    while (minHeap.size() > 1) {
        Node *left = minHeap.getMin();
        minHeap.removeMin();
        
        Node *right = minHeap.getMin();
        minHeap.removeMin();
        
        Node *temp = new Node(left->val + right->val,'\0');
        temp->l = left;
        temp->r = right;

        minHeap.insert(temp);
        
        // Set root to the last combined node
        root = temp;
    }
    return root;
    // unordered_map<char,string> huffmanCodes;
    // generateCodes(root,"",huffmanCodes);
    // printCodes(huffmanCodes);
    // string encoded = encode(str,huffmanCodes);
    // cout<<encoded<<endl;
    // decode(encoded,root);
}

Node* inputFunc(string str){
    unordered_map<char,int> input;
    
    int i=0;
    for(int i=0;i<str.size();i++){
        input[str[i]]+=1;
    }
    int size = input.size();
    char data[size];
    int val[size];
    for(const auto &pair:input){
        val[i] = pair.second;
        data[i] = pair.first;
        i++;
    }

    Node* root = huffman(str,data,val,size);
    return root;
}

int main() {
    string str = "This is huffman coding program";
    Node* root = inputFunc(str);
    // char data[] = {'A','B','C','D','E'};
    // int frequency[] = {5, 3, 8, 1, 6};

    // huffman(data,frequency, 5);
    // cout << endl;

    unordered_map<char,string> huffmanCodes;
    generateCodes(root,"",huffmanCodes);
    printCodes(huffmanCodes);
    string encoded = encode(str,huffmanCodes);
    cout<<encoded<<endl;
    decode(encoded,root);

    return 0;
}

File: huffman.py
class Node:
    def __init__(self, char, freq):
        self.char = char  # Character (None for internal nodes)
        self.freq = freq  # Frequency of the character
        self.left = None  # Left child (for internal nodes)
        self.right = None  # Right child (for internal nodes)


def build_huffman_tree(freq_map):
    # Step 1: Create leaf nodes from the frequency map
    nodes = []
    for char, freq in freq_map.items():
        nodes.append(Node(char, freq))

    # Step 2: Build the Huffman tree by merging nodes
    while len(nodes) > 1:
        # Sort the nodes based on frequency (ascending order)
        nodes.sort(key=lambda node: node.freq)

        # Take two nodes with the smallest frequencies
        left = nodes.pop(0)
        right = nodes.pop(0)

        # Merge two nodes into one
        merged = Node(None, left.freq + right.freq)  # New internal node with combined frequency
        merged.left = left
        merged.right = right

        # Add the merged node back to the list of nodes
        nodes.append(merged)

    # The last remaining node is the root of the Huffman tree
    return nodes[0]  # Return the root of the Huffman tree

def generate_codes(node, current_code, codes):
    if node is None:
        return

    if node.char is not None:  # If it's a leaf node
        codes[node.char] = current_code

    # Traverse left and right subtrees
    generate_codes(node.left, current_code + "0", codes)
    generate_codes(node.right, current_code + "1", codes)

def encode(input_string, huffman_codes):
    encoded_str = ''
    for char in input_string:
        encoded_str += huffman_codes[char]
    return encoded_str

def decode(encoded_str, root):
    decoded_str = ''
    current_node = root
    for bit in encoded_str:
        if bit == '0':
            current_node = current_node.left
        else:
            current_node = current_node.right

        if current_node.char is not None:  # Leaf node
            decoded_str += current_node.char
            current_node = root  # Reset to root for the next character

    return decoded_str

# Example usage:
input_string = str(input("enter string: "))

# Step 1: Calculate frequency of each character
freq_map = {}
for char in input_string:
    if char in freq_map:
        freq_map[char] += 1
    else:
        freq_map[char] = 1

# Step 2: Build Huffman Tree
root = build_huffman_tree(freq_map)

# Step 3: Generate Huffman Codes
huffman_codes = {}
generate_codes(root, "", huffman_codes)

# Step 4: Encode the input string
encoded_str = encode(input_string, huffman_codes)

# Step 5: Decode the encoded string
decoded_str = decode(encoded_str, root)
print("\nHuffman Codes:", huffman_codes)
print("Encoded String:", encoded_str)
print("Decoded String:", decoded_str)
